### PR TITLE
Added on cluster clause for clickhouse

### PR DESCRIPTION
**Pull request by diman-neosec**

### CODE REVIEW

**Comment by jahilliard** [comment]
Is this waiting on anything?

**Comment by diman-neosec** [comment]
> Is this waiting on anything?

Hi, we needed the ability to execute ClickHouse migrations on cluster.
So we have forked the repository.
How do you think it's best to add this ability if at all to this repo?
Can this PR be merged?

File: pkg/driver/clickhouse/clickhouse.go [hunk 1]
```diff
@@ -232,15 +249,25 @@ func (drv *Driver) DatabaseExists() (bool, error) {
...
 // CreateMigrationsTable creates the schema migrations table
 func (drv *Driver) CreateMigrationsTable(db *sql.DB) error {
+	onCluster := drv.onCluster()
+
+	onClusterClause := ""
+	engineClause := "engine = ReplacingMergeTree(ts)"
+
+	if onCluster {
+		onClusterClause = "on cluster '{cluster}'"
+		engineClause = "engine = ReplicatedReplacingMergeTree('/clickhouse/{installation}/{cluster}/tables/{table}', '{replica}', ts)"
```
Comment (amacneil) [inline pkg/driver/clickhouse/clickhouse.go:259]: what does this do? can you comment it?

File: pkg/driver/clickhouse/clickhouse.go [hunk 2]
```diff
@@ -96,6 +96,11 @@ func (drv *Driver) databaseName() string {
 	return name
 }
 
+func (drv *Driver) onCluster() bool {
+	onCluster := dbutil.MustParseURL(connectionString(drv.databaseURL)).Query().Get("on_cluster")
```
Comment (amacneil) [inline pkg/driver/clickhouse/clickhouse.go:100]: is the `on_cluster=true` URL parameter a standard ClickHouse connection string, or did you invent it?

File: pkg/driver/clickhouse/clickhouse.go [hunk 3]
```diff
@@ -119,7 +124,11 @@ func (drv *Driver) CreateDatabase() error {
 	}
 	defer dbutil.MustClose(db)
 
-	_, err = db.Exec("create database " + drv.quoteIdentifier(name))
+	if drv.onCluster() {
+		_, err = db.Exec("create database " + drv.quoteIdentifier(name) + " on cluster '{cluster}'")
```
Comment (amacneil) [inline pkg/driver/clickhouse/clickhouse.go:128]: Forgive me for not being familiar with ClickHouse, but why is the SQL statement here `on cluster '{cluster}'`? Don't you need to specify the cluster? Where does the cluster name come from?

**Review by amacneil — COMMENTED** [review]
Hi! Thanks for the PR. Before merging, can you address my inline questions. Also:

- Please add a PR description explaining the motivation behind this change and how to use it
- Please update the readme with documentation

Also pinging @Flamefork for a review since he contributed this code.

**Comment by amacneil** [comment]
Closing because no response to outstanding questions. I'm not familiar with ClickHouse but at the very least this PR needs a description explaining the change, and some tests. It also looks like it may contain a bug (see my comment above).

Please reopen if the above issues can be addressed.

**Comment by Flamefork** [comment]
Sorry, had noticed the ping only today.
I'm not sure what's the best way to do it, mostly cause I don't use ClickHouse's [Distributed DDL](https://clickhouse.com/docs/en/sql-reference/distributed-ddl/) feature and am not familiar with its potential caveats.

**Comment by diman-neosec** [comment]
Hey guys, thanks for all the replies, much appreciated.
Sorry I didn't have chance to push this forward.
You raised some great points, this solution is quite patchy. We need to think of a more robust way to support distributed ClickHouse migrations on dbmate.
Maybe a configuration file with tables engine and all other params requires.
In any case I think I will advice with CH community.

**Pull request closed without merging**